src/platform/PICO/uart/uart_pio.c

```diff
@@ -0,0 +1,374 @@
+        const int resourceIndex = serialResourceIndex(identifier);
+        const ioTag_t cfgRx = pSerialPinConfig->ioTagRx[resourceIndex];
+        const ioTag_t cfgTx = pSerialPinConfig->ioTagTx[resourceIndex];
+        bprintf("pico uartPinConfigure pio at %p dev = %p,  tags rx 0x%x, tx 0x%x", hardware, uartdev, cfgRx, cfgTx);
+        if (!cfgRx && !cfgTx) {
+            continue;
+        }
+
+        // On a single PIO block, we are restricted either to pins 0-31 or pins 16-47.
+        pinIndexMin = cfgRx && (DEFIO_TAG_PIN(cfgRx) < pinIndexMin) ? DEFIO_TAG_PIN(cfgRx) : pinIndexMin;
+        pinIndexMax = cfgRx && (DEFIO_TAG_PIN(cfgRx) > pinIndexMax) ? DEFIO_TAG_PIN(cfgRx) : pinIndexMax;
+        pinIndexMin = cfgTx && (DEFIO_TAG_PIN(cfgTx) < pinIndexMin) ? DEFIO_TAG_PIN(cfgTx) : pinIndexMin;
+        pinIndexMax = cfgTx && (DEFIO_TAG_PIN(cfgTx) > pinIndexMax) ? DEFIO_TAG_PIN(cfgTx) : pinIndexMax;
+        if (pinIndexMax >= 32) {
+            if (pinIndexMin < 16) {
+                bprintf("* Not configuring PIOUART with identifier %d (PIO can't span pins min %d max %d)",
+                        identifier, pinIndexMin, pinIndexMax);
+                continue;
+            } else {
+                uartPioBase = 16;
+            }
+        }
+
+        if (cfgRx) {
+            uartdev->rx = makePinDef(cfgRx);
+        }
+
+        if (cfgTx) {
+            uartdev->tx = makePinDef(cfgTx);
+        }
+
+        if (uartdev->rx.pin || uartdev->tx.pin ) {
+            bprintf("uartdev %p setting hardware to %p which has txbuffer %p", uartdev, hardware, hardware->txBuffer);
+            uartdev->hardware = (uartHardware_t *)hardware; // Sneak in pointer to pioUartHardware_t as a pointer to uartHardware_t
+        } else {
+            bprintf("** uartPinConfigure_pio no compatible rx or tx pin for this PIO UART");
+        }
+    }
+
+    bprintf("pico uartPinConfigure pio%d pin min, max = %d, %d; setting gpio base to %d", PIO_NUM(uartPio), pinIndexMin, pinIndexMax, uartPioBase);
+}
+
+static bool ensurePioProgram(PIO pio, const pio_program_t *program, bool isTx)
+{
+    // The GPIO base must be set before adding the program.
+    pio_set_gpio_base(uartPio, uartPioBase);
+
+    if (isTx) {
+        if (txProgram_offset < 0) {
+            txProgram_offset = pio_add_program(pio, program);
+        }
+
+        return txProgram_offset >= 0;
+   } else {
+        if (rxProgram_offset < 0) {
+            rxProgram_offset = pio_add_program(pio, program);
+        }
+
+        return rxProgram_offset >= 0;
+    }
+}
+
+#if SERIAL_PIOUART_COUNT > 0
+static void uartPioIrqHandler(uartPort_t *s, pioDetails_t *pioDetailsPtr)
+{
+    io_rw_32 *enableRegPtr = pioDetailsPtr->enableReg;
+    io_ro_32 *statusRegPtr = pioDetailsPtr->statusReg;
+    // Is RX enabled and does irq status indicate RX fifo not empty?
+    if (*enableRegPtr & pioDetailsPtr->rx_intr_bit && *statusRegPtr & pioDetailsPtr->rx_intr_bit) {
```

Comment:
Or get masked status directly. There is not need to read bits multiple times
```suggestion
    uint32_t pioIntrBits = *pioDetailsPtr->statusReg & *pioDetailsPtr->enableReg;
    // Is RX enabled and does irq status indicate RX fifo not empty?
    if (pioIntrBits & pioDetailsPtr->rx_intr_bit) {
```